### PR TITLE
Improve WASM Memory Ergonomics

### PR DESCRIPTION
**Pull request by ehennenfent**
Adds several helper methods to the WASM memory so as to conform to the standard laid out by native modules. Also lazily creates memory pages so that binaries can request large amounts of memory without Manticore needing to fill an appropriately large region with 0's.

### CODE REVIEW

File: manticore/wasm/structure.py [hunk 1]
```diff
@@ -442,7 +454,31 @@ def load(cls, filename: str):
...
-            # TODO - custom sections (https://www.w3.org/TR/wasm-core-1/#custom-section%E2%91%A0)
+            elif sec_id == SEC_UNK:
+                # WASM module renders all types as `GeneratedStructureData` so `isinstance` doesn't work :(
+                if (
+                    hasattr(section, "name_type")
+                    and hasattr(section, "payload_len")
+                    and hasattr(section, "payload")
+                ):
+                    name_type = section_data.name_type
+                    if name_type == 0:  # module name
```
Comment (ekilmer) [inline manticore/wasm/structure.py]: nit: Should we make an `enum` for the values of `name_type`?

File: manticore/wasm/structure.py [hunk 2]
```diff
@@ -442,7 +454,31 @@ def load(cls, filename: str):
...
+                            )
+                    elif name_type == 2:  # local variable names
+                        for func in section_data.payload.funcs:
+                            func_idx = func.index
+                            for n in func.local_map.names:
+                                ty = n.get_decoder_meta()["types"]["name_str"]
+                                m.local_names.setdefault(FuncAddr(func_idx), {})[
+                                    n.index
+                                ] = strip_quotes(ty.to_string(n.name_str))
+                # TODO - other custom sections (https://www.w3.org/TR/wasm-core-1/#custom-section%E2%91%A0)
```
Comment (ekilmer) [inline manticore/wasm/structure.py]: Unsupported sections should probably log a warning or debug message

File: manticore/wasm/structure.py [hunk 3]
```diff
@@ -488,8 +523,60 @@ class MemInst:
     https://www.w3.org/TR/wasm-core-1/#memory-instances%E2%91%A0
     """
 
-    data: typing.List[int]  #: The backing array for this memory
+    _published_events = {"write_memory", "read_memory"}
```
Comment (ekilmer) [inline manticore/wasm/structure.py]: I'm surprised `mypy` isn't unhappy. Usually it yells at me for missing class fields with type annotations.

Should be something like 

```suggestion
    _published_events: typing.Set[str] = {"write_memory", "read_memory"}
```

File: manticore/wasm/structure.py [hunk 4]
```diff
@@ -488,8 +523,60 @@ class MemInst:
...
+
+    _pages: typing.Dict[int, typing.List[int]]
     max: typing.Optional[U32]  #: Optional maximum number of pages the memory can contain
+    _current_size: int  # Tracks the theoretical size of the memory instance, including unmapped pages
+
+    def __init__(self, starting_data, max=None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        pagesize = 2 ** 16
```
Comment (ekilmer) [inline manticore/wasm/structure.py]: Can this value change? Should it be an exposed constant or class member?

File: manticore/wasm/structure.py [hunk 5]
```diff
@@ -500,16 +587,68 @@ def grow(self, n: int) -> bool:
...
+
+        :param base: Address to read the int from
+        :param size: Size of the int (in bits)
+        :return: The int in question
+        """
+        return Operators.CONCAT(
+            size, *map(Operators.ORD, reversed(self.read_bytes(base, size // 8)))
+        )
+
+    def read_bytes(self, base, size):
```
Comment (ekilmer) [inline manticore/wasm/structure.py]: Nit: Possible to add type annotations to these functions?

File: manticore/wasm/structure.py [hunk 6]
```diff
@@ -488,8 +523,60 @@ class MemInst:
...
+    def __contains__(self, item):
+        return item in range(self.npages * (2 ** 16))
+
+    def _check_initialize_index(self, memidx):
+        page = memidx // (2 ** 16)
+        if page not in range(self.npages):
+            raise OutOfBoundsMemoryTrap(memidx)
+        if page not in self._pages:
+            self._pages[page] = [0x0] * (2 ** 16)
+        return divmod(memidx, 2 ** 16)
```
Comment (ekilmer) [inline manticore/wasm/structure.py]: Looks like the `pagesize` value is used in a few places here again.

File: tests/wasm/generate_tests.sh [hunk 7]
```diff
@@ -40,4 +40,6 @@ chmod +x gen.sh
 cat modules.txt | xargs -n1 -P"$cores" ./gen.sh
 rm gen.sh
 
+mv test_callbacks.skip test_callbacks.py
```
Comment (ekilmer) [inline tests/wasm/generate_tests.sh]: Why is this needed?

Comment (ehennenfent) [reply tests/wasm/generate_tests.sh]: The WASM examples don't exhibit all the behaviors that Manticore can fire callbacks for, so `test_callbacks.skip` draws on examples from the spec tests for things like global variable usage, traps, and memory usage. Since those tests won't exist unless you've run generate_tests.sh, I disabled `test_callbacks` until you've done so by renaming the file.

Comment (ekilmer) [reply tests/wasm/generate_tests.sh]: Fair enough! My main concern was that the `test_callbacks.skip` might not be formatted or linted (if we ever eventually lint tests) by tools that look for the `*.py` extension.

File: manticore/wasm/structure.py (hunk 3 above)
Comment (ehennenfent) [reply manticore/wasm/structure.py]: It's because it inherits from Eventful, which already specifies the type.
`_published_events: Set[str] = set()`

File: manticore/wasm/structure.py (hunk 5 above)
Comment (ehennenfent) [reply manticore/wasm/structure.py]: Done. The type hints only describe concrete argument, which is fine since that's all we support right now. If we add a symbolic memory model we'll need to update them.

File: manticore/wasm/structure.py (hunk 1 above)
Comment (ehennenfent) [reply manticore/wasm/structure.py]: There are hardcoded constants in the `wasm` module, but it looks like they're out of date relative to the spec. I added a comment pointing to the spec itself; hopefully that will suffice.
